text/0082-combined-replay-envelope-item.md

```diff
@@ -0,0 +1,102 @@
+graph
+    A([SDK Generated Event])-->|HTTP Request| Z{Relay}
+    Z-->|Published to Kafka topic|B[Replays Kafka Cluster]
+    B-->|ingest-replay-events|J[Snuba Consumer]
+    J-->|Buffered Bulk Commit|K[(Clickhouse)]
+    B-->|ingest-replay-recordings|E[Recording Consumer]
+    E-->C[Recording Chunk Processor]
+    C-->|Chunk Stashed|D[(Redis)]
+    E-->Q[Recording Event Processor]
+    Q-->F[Chunk Assembly]
+    I[(Redis)]-->|Chunks Fetched|F
+    F-->|Assembled Chunks Stored|G[(Blob Storage)]
+    F-->|File Metadata Saved|H[(PostgreSQL)]
+```
+
+We'd like to combine these envelope payloads into one for the following reasons:
+
+- Now that we're decompressing, parsing, and indexing data in the recording, the delineation between the two events no longer makes sense.
+- Right now there exists a race condition between ReplayEvent and ReplayRecording -- if a ReplayEvent makes it to clickhouse and is stored before the ReplayRecording, it can result in a bad user experience as a user can navigate to the replay, but the replay's blobs may not be stored yet. We'd like to change it so the snuba writes happen _downstream_ from the recording consumer, as we don't want to make a replay available for search until it's corresponding recording blob has been stored.
+- Right now our rate limits are separated per Itemtype. It would be less confusing if the rate limit applied to a single event type.
+- It is very hard to do telemetry on the recordings consumer now as we do not have SDK version. combining the envelopes allows us to have metadata in our recordings consumer in an easily accessible way.
+
+# Options Considered
+
+New Proposed Flow:
+
+```mermaid
+graph
+    A([SDK Generated Combined Replay Envelope Item])--> Z[Relay, Combines Envelope Items if needed]
+    AA([Legacy SDK Generated Separate Items]) --> Z
+    Z-->|Published to Kafka topic, using random partition key|B[Replays Kafka Cluster]
+    T-->|ingest-replay-events|J[Snuba Consumer]
+    J-->|Buffered Bulk Commit|K[(Clickhouse)]
+    B-->|ingest-replay-recordings|E[Recording Consumer]
+    E-->Q[Parse Replay for Click Events, etc.]
+    Q-->R[Emit Outcome if segment = 0]
+    Q-->S[Store Replay Blob in Storage, GCS in SaaS ]
+    S-->T[ Emit Kafka Message for Snuba Consumer]
+```
+
+### Relay Changes:
+
+1. Create a new ItemType `CombinedReplayRecordingEvent`
+2. in `processor.rs` https://github.com/getsentry/relay/blob/606166fca57a84ca1b9240253013871d13827de3/relay-server/src/actors/processor.rs#L1134, replace the `process_replays` function that will combine the envelope items if its the old format using `take_item_by` and `add_item` Envelope functions, or leave them if its the new format. This function would then process the two items and set the item payload.
+3. We can begin rejecting replay segments over 10 megabytes, which will allow us to not need chunking in our recording consumer.
+
+- The combined Payload will have the format of
+
+```
+ReplayEventJSON\nCompressedReplayRecording
```

Comment:
Instead of making this a newline-separated string, could we also send this as a JSON like this:
```
{
  event: ReplayEventJSON,
  recording: CompressedReplayRecording
}
```

Not sure about implications for parsing this, but from an SDK perspective this seems a bit nicer/safer than string-concatting this. But not a hard feeling, just a thought.

Reply:
a wait, I guess we can't put the compressed stuff into JSON 🤔 but now that I think about this, not sure if we can concat the UInt8Array that is the compressed payload onto a string as well 🤔

Reply:
If this merged type is generated on the client then you will need to serialize with a binary format like `msgpack`.

Reply:
hmm, I see. I guess we can do that, convert the payload into UInt8Array if it isn't compressed, then combine them together into a single UInt8Array.

One downside to note here is that it may make debugging harder - I currently tend to disable compression when debugging issues, in order to see what data is sent. This would not be possible anymore (as the payload would then _always_ be UInt8Array). Not a blocker, just to be aware of this downside!